File: android/app/src/main/java/com/hedvig/app/ApplicationModule.java
package com.hedvig.app;

import android.app.Application;
import android.content.Context;

import com.RNFetchBlob.RNFetchBlobPackage;
import com.airbnb.android.react.lottie.LottiePackage;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.Logger;
import com.apollographql.apollo.internal.ApolloLogger;
import com.brentvatne.react.ReactVideoPackage;
import com.dylanvann.fastimage.FastImageViewPackage;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactPackage;
import com.facebook.react.common.LifecycleState;
import com.facebook.react.shell.MainReactPackage;
import com.google.android.exoplayer2.upstream.cache.LeastRecentlyUsedCacheEvictor;
import com.google.android.exoplayer2.upstream.cache.SimpleCache;
import com.hedvig.android.owldroid.util.react.AsyncStorageNativeReader;
import com.hedvig.app.starter.ActivityStarterReactPackage;
import com.horcrux.svg.SvgPackage;
import com.imagepicker.ImagePickerPackage;
import com.learnium.RNDeviceInfo.RNDeviceInfo;
import com.leo_pharma.analytics.AnalyticsPackage;
import com.lugg.ReactNativeConfig.ReactNativeConfigPackage;
import com.reactnativedocumentpicker.ReactNativeDocumentPicker;
import com.rnfs.RNFSPackage;
import com.rnim.rn.audio.ReactNativeAudioPackage;
import com.swmansion.gesturehandler.react.RNGestureHandlerPackage;
import com.zmxv.RNSound.RNSoundPackage;

import java.io.File;
import java.util.Arrays;
import java.util.List;

import javax.annotation.Nullable;
import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.branch.rnbranch.RNBranchPackage;
import io.invertase.firebase.RNFirebasePackage;
import io.invertase.firebase.analytics.RNFirebaseAnalyticsPackage;
import io.invertase.firebase.messaging.RNFirebaseMessagingPackage;
import io.invertase.firebase.notifications.RNFirebaseNotificationsPackage;
import io.sentry.RNSentryPackage;
import okhttp3.logging.HttpLoggingInterceptor;

@Module
public class ApplicationModule {
    @Provides
    static Application application(MainApplication application) {
        return application;
    }

    @Provides
    static Context context(Application application) {
        return application.getBaseContext();
    }

    @Provides
    static SimpleCache simpleCache(Context context) {
        return new SimpleCache(new File(context.getCacheDir(), "hedvig_story_video_cache"), new LeastRecentlyUsedCacheEvictor(10 * 1024 * 1024));
    }

    @Provides
    static AsyncStorageNativeReader asyncStorageNativeReader(Context context) {
        return new AsyncStorageNativeReaderImpl(context);
    }

    @Provides
    @Named("GRAPHQL_URL")
    static String graphqlUrl() {
        return BuildConfig.GRAPHQL_URL;
    }

    @Provides
    @Named("VERSION_NUMBER")
    static String versionNumber() {
        return BuildConfig.VERSION_NAME;
    }

    @Provides
    @Named("APPLICATION_ID")
    static String applicationId() {
        return BuildConfig.APPLICATION_ID;
    }

    @Provides
    @Nullable
    static Logger apolloLogger() {
        return null;
    }

    @Provides
    @Nullable
    static HttpLoggingInterceptor httpLoggingInterceptor() {
        return null;
    }

    @Provides
    @Singleton
    static ReactInstanceManager reactInstanceManager(MainApplication application, ApolloClient apolloClient) {
        return ReactInstanceManager
                .builder()
                .setApplication(application)
                .addPackages(getPackages(apolloClient))
                .setJSMainModulePath("androidtest")
                .setInitialLifecycleState(LifecycleState.RESUMED)
                .setUseDeveloperSupport(true)
                .build();
    }

    static List<ReactPackage> getPackages(ApolloClient apolloClient) {
        return Arrays.asList(
                new ActivityStarterReactPackage(),
                new MainReactPackage(),
                new ReactNativeDocumentPicker(),
                new FastImageViewPackage(),
                new ImagePickerPackage(),
                new RNFSPackage(),
                new ReactVideoPackage(),
                new RNGestureHandlerPackage(),
                new RNDeviceInfo(),
                new SvgPackage(),
                new ReactNativeConfigPackage(),
                new RNFetchBlobPackage(),
                new RNSoundPackage(),
                new RNSentryPackage(),
                new RNFirebasePackage(),
                new RNFirebaseNotificationsPackage(),
                new RNFirebaseMessagingPackage(),
                new RNFirebaseAnalyticsPackage(),
                new RNBranchPackage(),
                new ReactNativeAudioPackage(),
                new AnalyticsPackage(),
                new LottiePackage(),
                new NativeRoutingPackage(apolloClient)
        );
    }
}

File: android/app/src/main/java/com/hedvig/app/di/ApplicationModule.kt
package com.hedvig.app.di

import android.app.Application
import android.content.Context
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.Logger
import com.apollographql.apollo.cache.normalized.NormalizedCacheFactory
import com.apollographql.apollo.cache.normalized.lru.EvictionPolicy
import com.apollographql.apollo.cache.normalized.lru.LruNormalizedCache
import com.apollographql.apollo.cache.normalized.lru.LruNormalizedCacheFactory
import com.google.android.exoplayer2.upstream.cache.LeastRecentlyUsedCacheEvictor
import com.google.android.exoplayer2.upstream.cache.SimpleCache
import com.google.firebase.analytics.FirebaseAnalytics
import com.hedvig.android.owldroid.type.CustomType
import com.hedvig.app.BuildConfig
import com.hedvig.app.MainApplication
import com.hedvig.app.util.apollo.ApolloTimberLogger
import com.hedvig.app.util.apollo.PromiscuousLocalDateAdapter
import com.hedvig.app.util.react.AsyncStorageNative
import com.hedvig.app.util.react.AsyncStorageNativeImpl
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import timber.log.Timber
import java.io.File
import javax.inject.Named
import javax.inject.Singleton

@Module
object ApplicationModule {
    @Provides
    @JvmStatic
    fun application(application: MainApplication): Application = application

    @Provides
    @JvmStatic
    fun context(application: Application) = application.baseContext

    @Provides
    @JvmStatic
    fun firebaseAnalytics(context: Context) = FirebaseAnalytics.getInstance(context)

    @Provides
    @JvmStatic
    fun simpleCache(context: Context) = SimpleCache(
        File(context.cacheDir, "hedvig_story_video_cache"),
        LeastRecentlyUsedCacheEvictor((10 * 1024 * 1024).toLong())
    )

    @Provides
    @JvmStatic
    fun asyncStorageNativeReader(context: Context): AsyncStorageNative =
        AsyncStorageNativeImpl(context)


    @Provides
    @JvmStatic
    fun apolloLogger(): Logger? = if (BuildConfig.DEBUG || BuildConfig.APP_ID == "com.hedvig.test.app") {
        ApolloTimberLogger()
    } else {
        null
    }

    @Provides
    @JvmStatic
    fun httpLoggingInterceptor(): HttpLoggingInterceptor? =
        if (BuildConfig.DEBUG || BuildConfig.APP_ID == "com.hedvig.test.app") {
            HttpLoggingInterceptor(HttpLoggingInterceptor.Logger { message ->
                Timber.tag("OkHttp").i(message)
            }).setLevel(HttpLoggingInterceptor.Level.BODY)
        } else {
            null
        }

    @Provides
    @Singleton
    @JvmStatic
    fun okHttpClient(
        asyncStorageNative: AsyncStorageNative,
        httpLoggingInterceptor: HttpLoggingInterceptor?
    ): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .addInterceptor { chain ->
                val original = chain.request()
                val builder = original
                    .newBuilder()
                    .method(original.method(), original.body())
                try {
                    asyncStorageNative.getKey("@hedvig:token")
                } catch (exception: Exception) {
                    Timber.e(exception, "Got an exception while trying to retrieve token")
                    null
                }?.let { token ->
                    builder.header("Authorization", token)
                }
                chain.proceed(builder.build())
            }
        httpLoggingInterceptor?.let { builder.addInterceptor(it) }
        return builder.build()
    }

    @Provides
    @Singleton
    @JvmStatic
    fun lruNormalizedCacheFactory(): NormalizedCacheFactory<LruNormalizedCache> {
        return LruNormalizedCacheFactory(EvictionPolicy.builder().maxSizeBytes(1000 * 1024).build())
    }

    @Provides
    @Singleton
    @JvmStatic
    fun apolloClient(
        okHttpClient: OkHttpClient,
        normalizedCacheFactory: NormalizedCacheFactory<LruNormalizedCache>,
        @Named("GRAPHQL_URL") graphqlUrl: String,
        logger: Logger?
    ): ApolloClient {
        val builder = ApolloClient
            .builder()
            .serverUrl(graphqlUrl)
            .okHttpClient(okHttpClient)
            .addCustomTypeAdapter(CustomType.LOCALDATE, PromiscuousLocalDateAdapter())
            .normalizedCache(normalizedCacheFactory)

        logger?.let { builder.logger(it) }

        return builder.build()
    }
}
